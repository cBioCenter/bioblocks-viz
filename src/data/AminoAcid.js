"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var AminoAcid = /** @class */ (function () {
    function AminoAcid(fullName, singleLetterCode, threeLetterCode, codons) {
        this.fullName = fullName;
        this.singleLetterCode = singleLetterCode;
        this.threeLetterCode = threeLetterCode;
        this.codons = codons;
    }
    // public static readonly Stop = new AminoAcid('Stop', '*', '***', ['TAA', 'TGA', 'TAG']);
    // public static readonly Gap = new AminoAcid('Gap', '-', '---', []);
    // public static readonly Unknown = new AminoAcid('Unknown', '?', '???', []);
    AminoAcid.getAllCanonicalAminoAcids = function () {
        return Object.keys(AminoAcid).reduce(function (acc, staticVarName) {
            var staticValue = AminoAcid[staticVarName];
            if (staticValue instanceof AminoAcid) {
                acc.push(staticValue);
            }
            return acc;
        }, new Array());
    };
    AminoAcid.fromSingleLetterCode = function (code) {
        return AminoAcid.getAllCanonicalAminoAcids().find(function (aa) {
            if (aa.singleLetterCode.toUpperCase() === code.toUpperCase()) {
                return true;
            }
            return false;
        });
    };
    AminoAcid.fromThreeLetterCode = function (code) {
        AminoAcid.getAllCanonicalAminoAcids().forEach(function (aa) {
            if (aa.threeLetterCode.toUpperCase() === code.toUpperCase()) {
                return aa;
            }
        });
        return undefined;
    };
    AminoAcid.fromCodon = function (codon) {
        AminoAcid.getAllCanonicalAminoAcids().forEach(function (aa) {
            aa.codons.forEach(function (codonStr) {
                if (codonStr.toUpperCase() === codon.toUpperCase()) {
                    return aa;
                }
            });
        });
        return undefined;
    };
    AminoAcid.Alanine = new AminoAcid('Alanine', 'A', 'ALA', ['GCT', 'GCC', 'GCA', 'GCG']);
    AminoAcid.Arginine = new AminoAcid('Arginine', 'R', 'ARG', ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG']);
    AminoAcid.Asparagine = new AminoAcid('Asparagine', 'N', 'ASN', ['AAT', 'AAC']);
    AminoAcid.AsparticAcid = new AminoAcid('Aspartic Acid', 'D', 'ASP', ['GAT', 'GAC']);
    AminoAcid.Cysteine = new AminoAcid('Cysteine', 'C', 'CYS', ['TGT', 'TGC']);
    AminoAcid.Glutamine = new AminoAcid('Glutamine', 'Q', 'GLN', ['CAA', 'CAG']);
    AminoAcid.GlutamicAcid = new AminoAcid('Glutamic Acid', 'E', 'GLU', ['GAA', 'GAG']);
    AminoAcid.Glycine = new AminoAcid('Glycine', 'G', 'GLY', ['GGT', 'GGC', 'GGA', 'GGG']);
    AminoAcid.Histidine = new AminoAcid('Histidine', 'H', 'HIS', ['CAT', 'CAC']);
    AminoAcid.Isoleucine = new AminoAcid('Isoleucine', 'I', 'ILE', ['ATT', 'ATC', 'ATA']);
    AminoAcid.Leucine = new AminoAcid('Leucine', 'L', 'LEU', ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG']);
    AminoAcid.Lysine = new AminoAcid('Lysine', 'K', 'LYS', ['AAA', 'AAG']);
    AminoAcid.Methionine = new AminoAcid('Methionine', 'M', 'MET', ['ATG']);
    AminoAcid.Phenylalanine = new AminoAcid('Phenylalanine', 'F', 'PHE', ['TTT', 'TTC']);
    AminoAcid.Proline = new AminoAcid('Proline', 'P', 'PRO', ['CCT', 'CCC', 'CCA', 'CCG']);
    AminoAcid.Serine = new AminoAcid('Serine', 'S', 'SER', ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC']);
    AminoAcid.Threonine = new AminoAcid('Threonine', 'T', 'THR', ['ACT', 'ACC', 'ACA', 'ACG']);
    AminoAcid.Tryptophan = new AminoAcid('Tryptophan', 'W', 'TRP', ['TGG']);
    AminoAcid.Tyrosine = new AminoAcid('Tyrosine', 'Y', 'TYR', ['TAT', 'TAC']);
    AminoAcid.Valine = new AminoAcid('Valine', 'V', 'VAL', ['GTT', 'GTC', 'GTA', 'GTG']);
    // public static readonly Unknown = new AminoAcid('Unknown', 'X', 'XXX', []);
    AminoAcid.ALL_FULL_NAMES = function () { return AminoAcid.getAllCanonicalAminoAcids().map(function (aa) { return aa.fullName; }); };
    AminoAcid.ALL_1LETTER_CODES = function () { return AminoAcid.getAllCanonicalAminoAcids().map(function (a) { return a.singleLetterCode; }); };
    AminoAcid.ALL_3LETTER_CODES = function () { return AminoAcid.getAllCanonicalAminoAcids().map(function (a) { return a.threeLetterCode; }); };
    AminoAcid.ALL_CODONS = function () { return lodash_1.flatten(AminoAcid.getAllCanonicalAminoAcids().map(function (a) { return a.codons; })); };
    return AminoAcid;
}());
exports.AminoAcid = AminoAcid;
// Gets mismatches between a PDB file and coupling scores.
// @returns Array of mismatches - empty if none are found.
exports.getPDBAndCouplingMismatch = function (pdbData, couplingScores) {
    var pdbSequence = pdbData.sequence;
    var couplingSequence = couplingScores.sequence;
    var mismatches = exports.getSequenceMismatch(pdbSequence, couplingSequence);
    if (mismatches.length === 0) {
        return pdbData.getResidueNumberingMismatches(couplingScores);
    }
    return mismatches;
};
// Given two **equal-length** Amino Acid sequences, returns all mismatches.
exports.getSequenceMismatch = function (firstSequence, secondSequence) {
    var mismatches = new Array();
    if (firstSequence.length === secondSequence.length) {
        for (var i = 0; i < firstSequence.length; ++i) {
            var couplingAminoAcid = AminoAcid.fromSingleLetterCode(firstSequence[i]);
            var pdbAminoAcid = AminoAcid.fromSingleLetterCode(secondSequence[i]);
            if (firstSequence[i] !== secondSequence[i] && couplingAminoAcid && pdbAminoAcid) {
                mismatches.push({
                    couplingAminoAcid: couplingAminoAcid,
                    pdbAminoAcid: pdbAminoAcid,
                    resno: i,
                });
            }
        }
    }
    return mismatches;
};
//# sourceMappingURL=AminoAcid.js.map