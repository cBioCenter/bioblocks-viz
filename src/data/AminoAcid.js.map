{"version":3,"file":"AminoAcid.js","sourceRoot":"","sources":["../../../src/data/AminoAcid.ts"],"names":[],"mappings":";;AAAA,iCAAiC;AAQjC;IA6EE,mBACW,QAAgB,EAChB,gBAAwB,EACxB,eAAuB,EACvB,MAAyB;QAHzB,aAAQ,GAAR,QAAQ,CAAQ;QAChB,qBAAgB,GAAhB,gBAAgB,CAAQ;QACxB,oBAAe,GAAf,eAAe,CAAQ;QACvB,WAAM,GAAN,MAAM,CAAmB;IACjC,CAAC;IArDJ,0FAA0F;IAC1F,qEAAqE;IACrE,6EAA6E;IAE/D,mCAAyB,GAAvC;QACE,OAAO,MAAM,CAAC,IAAI,CAAC,SAAS,CAAC,CAAC,MAAM,CAAC,UAAC,GAAG,EAAE,aAAa;YACtD,IAAM,WAAW,GAAG,SAAS,CAAC,aAAuC,CAAC,CAAC;YAEvE,IAAI,WAAW,YAAY,SAAS,EAAE;gBACpC,GAAG,CAAC,IAAI,CAAC,WAAW,CAAC,CAAC;aACvB;YAED,OAAO,GAAG,CAAC;QACb,CAAC,EAAE,IAAI,KAAK,EAAa,CAAC,CAAC;IAC7B,CAAC;IAEa,8BAAoB,GAAlC,UAAmC,IAAY;QAC7C,OAAO,SAAS,CAAC,yBAAyB,EAAE,CAAC,IAAI,CAAC,UAAA,EAAE;YAClD,IAAI,EAAE,CAAC,gBAAgB,CAAC,WAAW,EAAE,KAAK,IAAI,CAAC,WAAW,EAAE,EAAE;gBAC5D,OAAO,IAAI,CAAC;aACb;YAED,OAAO,KAAK,CAAC;QACf,CAAC,CAAC,CAAC;IACL,CAAC;IAEa,6BAAmB,GAAjC,UAAkC,IAAY;QAC5C,SAAS,CAAC,yBAAyB,EAAE,CAAC,OAAO,CAAC,UAAA,EAAE;YAC9C,IAAI,EAAE,CAAC,eAAe,CAAC,WAAW,EAAE,KAAK,IAAI,CAAC,WAAW,EAAE,EAAE;gBAC3D,OAAO,EAAE,CAAC;aACX;QACH,CAAC,CAAC,CAAC;QAEH,OAAO,SAAS,CAAC;IACnB,CAAC;IAEa,mBAAS,GAAvB,UAAwB,KAAa;QACnC,SAAS,CAAC,yBAAyB,EAAE,CAAC,OAAO,CAAC,UAAA,EAAE;YAC9C,EAAE,CAAC,MAAM,CAAC,OAAO,CAAC,UAAC,QAAgB;gBACjC,IAAI,QAAQ,CAAC,WAAW,EAAE,KAAK,KAAK,CAAC,WAAW,EAAE,EAAE;oBAClD,OAAO,EAAE,CAAC;iBACX;YACH,CAAC,CAAC,CAAC;QACL,CAAC,CAAC,CAAC;QAEH,OAAO,SAAS,CAAC;IACnB,CAAC;IA1EsB,iBAAO,GAAG,IAAI,SAAS,CAAC,SAAS,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7E,kBAAQ,GAAG,IAAI,SAAS,CAAC,UAAU,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7F,oBAAU,GAAG,IAAI,SAAS,CAAC,YAAY,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACrE,sBAAY,GAAG,IAAI,SAAS,CAAC,eAAe,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC1E,kBAAQ,GAAG,IAAI,SAAS,CAAC,UAAU,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACjE,mBAAS,GAAG,IAAI,SAAS,CAAC,WAAW,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACnE,sBAAY,GAAG,IAAI,SAAS,CAAC,eAAe,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC1E,iBAAO,GAAG,IAAI,SAAS,CAAC,SAAS,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7E,mBAAS,GAAG,IAAI,SAAS,CAAC,WAAW,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACnE,oBAAU,GAAG,IAAI,SAAS,CAAC,YAAY,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC5E,iBAAO,GAAG,IAAI,SAAS,CAAC,SAAS,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC3F,gBAAM,GAAG,IAAI,SAAS,CAAC,QAAQ,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7D,oBAAU,GAAG,IAAI,SAAS,CAAC,YAAY,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,CAAC,CAAC,CAAC;IAC9D,uBAAa,GAAG,IAAI,SAAS,CAAC,eAAe,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC3E,iBAAO,GAAG,IAAI,SAAS,CAAC,SAAS,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7E,gBAAM,GAAG,IAAI,SAAS,CAAC,QAAQ,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACzF,mBAAS,GAAG,IAAI,SAAS,CAAC,WAAW,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACjF,oBAAU,GAAG,IAAI,SAAS,CAAC,YAAY,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,CAAC,CAAC,CAAC;IAC9D,kBAAQ,GAAG,IAAI,SAAS,CAAC,UAAU,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACjE,gBAAM,GAAG,IAAI,SAAS,CAAC,QAAQ,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAElG,6EAA6E;IAE/D,wBAAc,GAAG,cAAM,OAAA,SAAS,CAAC,yBAAyB,EAAE,CAAC,GAAG,CAAC,UAAA,EAAE,IAAI,OAAA,EAAE,CAAC,QAAQ,EAAX,CAAW,CAAC,EAA5D,CAA4D,CAAC;IACpF,2BAAiB,GAAG,cAAM,OAAA,SAAS,CAAC,yBAAyB,EAAE,CAAC,GAAG,CAAC,UAAA,CAAC,IAAI,OAAA,CAAC,CAAC,gBAAgB,EAAlB,CAAkB,CAAC,EAAlE,CAAkE,CAAC;IAC7F,2BAAiB,GAAG,cAAM,OAAA,SAAS,CAAC,yBAAyB,EAAE,CAAC,GAAG,CAAC,UAAA,CAAC,IAAI,OAAA,CAAC,CAAC,eAAe,EAAjB,CAAiB,CAAC,EAAjE,CAAiE,CAAC;IAC5F,oBAAU,GAAG,cAAM,OAAA,gBAAO,CAAC,SAAS,CAAC,yBAAyB,EAAE,CAAC,GAAG,CAAC,UAAA,CAAC,IAAI,OAAA,CAAC,CAAC,MAAM,EAAR,CAAQ,CAAC,CAAC,EAAjE,CAAiE,CAAC;IAwDrG,gBAAC;CAAA,AAnFD,IAmFC;AAnFY,8BAAS;AA6btB,0DAA0D;AAC1D,0DAA0D;AAC7C,QAAA,yBAAyB,GAAG,UAAC,OAAqB,EAAE,cAAiC;IAChG,IAAM,WAAW,GAAG,OAAO,CAAC,QAAQ,CAAC;IACrC,IAAM,gBAAgB,GAAG,cAAc,CAAC,QAAQ,CAAC;IACjD,IAAM,UAAU,GAAG,2BAAmB,CAAC,WAAW,EAAE,gBAAgB,CAAC,CAAC;IAEtE,IAAI,UAAU,CAAC,MAAM,KAAK,CAAC,EAAE;QAC3B,OAAO,OAAO,CAAC,6BAA6B,CAAC,cAAc,CAAC,CAAC;KAC9D;IAED,OAAO,UAAU,CAAC;AACpB,CAAC,CAAC;AAEF,2EAA2E;AAC9D,QAAA,mBAAmB,GAAG,UAAC,aAAqB,EAAE,cAAsB;IAC/E,IAAM,UAAU,GAAG,IAAI,KAAK,EAA0B,CAAC;IACvD,IAAI,aAAa,CAAC,MAAM,KAAK,cAAc,CAAC,MAAM,EAAE;QAClD,KAAK,IAAI,CAAC,GAAG,CAAC,EAAE,CAAC,GAAG,aAAa,CAAC,MAAM,EAAE,EAAE,CAAC,EAAE;YAC7C,IAAM,iBAAiB,GAAG,SAAS,CAAC,oBAAoB,CAAC,aAAa,CAAC,CAAC,CAAC,CAAC,CAAC;YAC3E,IAAM,YAAY,GAAG,SAAS,CAAC,oBAAoB,CAAC,cAAc,CAAC,CAAC,CAAC,CAAC,CAAC;YACvE,IAAI,aAAa,CAAC,CAAC,CAAC,KAAK,cAAc,CAAC,CAAC,CAAC,IAAI,iBAAiB,IAAI,YAAY,EAAE;gBAC/E,UAAU,CAAC,IAAI,CAAC;oBACd,iBAAiB,mBAAA;oBACjB,YAAY,cAAA;oBACZ,KAAK,EAAE,CAAC;iBACT,CAAC,CAAC;aACJ;SACF;KACF;IAED,OAAO,UAAU,CAAC;AACpB,CAAC,CAAC"}