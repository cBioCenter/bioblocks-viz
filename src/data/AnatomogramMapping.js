"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnatomogramMapping = {
    homo_sapiens: {
        P11A: ['UBERON_0000178'],
        P11B: ['UBERON_0001155'],
        P12A: ['UBERON_0000955'],
        P9A: ['UBERON_0000977'],
        UBERON_0000178: ['P11A'],
        UBERON_0000955: ['P12A'],
        UBERON_0000977: ['P9A'],
        UBERON_0001155: ['P11B'],
    },
    mus_musculus: {
        UBERON_0000948: ['Heart', 'Heart_and_Aorta', 'Heart_and_Aorta-10X_P7_4'],
        UBERON_0001255: ['Bladder', 'Bladder-10X_P4_3', 'Bladder-10X_P4_4', 'Bladder-10X_P7_7'],
        UBERON_0001723: ['Tongue', 'Tongue-10X_P4_0', 'Tongue-10X_P4_1', 'Tongue-10X_P7_10'],
        UBERON_0001911: ['Mammary', 'Mammary_Gland', 'Mammary_Gland-10X_P7_12', 'Mammary_Gland-10X_P7_13'],
        UBERON_0002048: ['Lung', 'Lung-10X_P7_8', 'Lung-10X_P7_9', 'Lung-10X_P8_12', 'Lung-10X_P8_13'],
        UBERON_0002106: ['Spleen', 'Spleen-10X_P4_7', 'Spleen-10X_P7_6'],
        UBERON_0002107: ['Liver', 'Liver-10X_P4_2', 'Liver-10X_P7_0', 'Liver-10X_P7_1'],
        UBERON_0002113: ['Kidney', 'Kidney-10X_P4_5', 'Kidney-10X_P4_6', 'Kidney-10X_P7_5'],
        UBERON_0002370: ['Thymus', 'Thymus-10X_P7_11'],
        UBERON_0002371: ['Marrow', 'Marrow-10X_P7_2', 'Marrow-10X_P7_3'],
        UBERON_0003126: ['Trachea', 'Trachea-10X_P8_14', 'Trachea-10X_P8_15'],
        UBERON_0014892: ['Muscle', 'Limb', 'Limb_Muscle', 'Limb_Muscle-10X_P7_14', 'Limb_Muscle-10X_P7_15'],
    },
};
//# sourceMappingURL=AnatomogramMapping.js.map