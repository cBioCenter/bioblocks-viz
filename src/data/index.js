"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./AminoAcid"), exports);
tslib_1.__exportStar(require("./Nucleotide"), exports);
tslib_1.__exportStar(require("./AnatomogramMapping"), exports);
tslib_1.__exportStar(require("./Bioblocks1DSection"), exports);
tslib_1.__exportStar(require("./BioblocksConfig"), exports);
tslib_1.__exportStar(require("./BioblocksData"), exports);
tslib_1.__exportStar(require("./BioblocksPDB"), exports);
tslib_1.__exportStar(require("./CodonTable"), exports);
tslib_1.__exportStar(require("./CouplingContainer"), exports);
tslib_1.__exportStar(require("./Dataset"), exports);
tslib_1.__exportStar(require("./IUPACData"), exports);
tslib_1.__exportStar(require("./Markers"), exports);
tslib_1.__exportStar(require("./PfamProtein"), exports);
tslib_1.__exportStar(require("./PlotlyData"), exports);
tslib_1.__exportStar(require("./Protein"), exports);
tslib_1.__exportStar(require("./Seq"), exports);
tslib_1.__exportStar(require("./SeqIO"), exports);
tslib_1.__exportStar(require("./SeqRecord"), exports);
tslib_1.__exportStar(require("./TintedBioblocks1DSection"), exports);
tslib_1.__exportStar(require("./event"), exports);
//# sourceMappingURL=index.js.map