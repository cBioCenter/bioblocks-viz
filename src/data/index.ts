export * from './AminoAcid';
export * from './Chell1DSection';
export * from './ChellConfig';
export * from './ChellData';
export * from './ChellPDB';
export * from './CouplingContainer';
export * from './PlotlyData';
export * from './Protein';
export * from './Spring';
export * from './TintedChell1DSection';

export * from './event';
