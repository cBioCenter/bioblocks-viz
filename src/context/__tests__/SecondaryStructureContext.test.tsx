import { mount } from 'enzyme';
import * as React from 'react';

import {
  initialSecondaryStructureContext,
  ISecondaryStructureContext,
  SecondaryStructureContextConsumer,
  SecondaryStructureContextProvider,
} from '~chell-viz~/context';
import { Chell1DSection, SECONDARY_STRUCTURE_KEYS } from '~chell-viz~/data';
import { genSecondaryStructureSection, MockContextClass } from '~chell-viz~/test';

describe('SecondaryStructureContext', () => {
  interface IMockContextConsumerProps {
    secondaryStructureContext: ISecondaryStructureContext;
  }

  let MockWrappedElement: JSX.Element;
  let MockContextConsumer: JSX.Element;
  let MockContextProvider: JSX.Element;
  beforeEach(() => {
    MockContextConsumer = (
      <SecondaryStructureContextConsumer>
        {secondaryStructureContext => {
          MockWrappedElement = <MockContextClass secondaryStructureContext={secondaryStructureContext} />;

          return MockWrappedElement;
        }}
      </SecondaryStructureContextConsumer>
    );
    MockContextProvider = <SecondaryStructureContextProvider>{MockContextConsumer}</SecondaryStructureContextProvider>;
  });

  describe('Consumer', () => {
    it('Should allow components to consume a secondary structure context.', () => {
      const wrapper = mount(MockContextConsumer);
      expect(wrapper).toMatchSnapshot();
      const props = wrapper.props() as IMockContextConsumerProps;
      expect(props.secondaryStructureContext).toEqual(initialSecondaryStructureContext);
    });

    it('Should have default implementations if no provider is provided.', () => {
      mount(MockContextConsumer);
      const context = (MockWrappedElement.props as IMockContextConsumerProps).secondaryStructureContext;
      const sectionToAdd = new Chell1DSection('B' as SECONDARY_STRUCTURE_KEYS, 0, 5);

      context.addSelectedSecondaryStructure(sectionToAdd);
      context.addHoveredSecondaryStructure(sectionToAdd);
      expect(context.selectedSecondaryStructures).toEqual([]);
      expect(context.hoveredSecondaryStructures).toEqual([]);
      context.removeHoveredSecondaryStructure(sectionToAdd);
      context.removeSecondaryStructure(sectionToAdd);
      context.clearAllSecondaryStructures();
      expect(context.selectedSecondaryStructures).toEqual([]);
      expect(context.hoveredSecondaryStructures).toEqual([]);
    });
  });

  describe('Provider', () => {
    it('Should allow components to be provided a secondary structure context.', () => {
      const wrapper = mount(MockContextProvider);
      expect(wrapper).toMatchSnapshot();
    });

    it('Should handle adding a selected secondary structure.', () => {
      const wrapper = mount(MockContextProvider);
      const context = (MockWrappedElement.props as IMockContextConsumerProps).secondaryStructureContext;
      const instance = wrapper.instance() as SecondaryStructureContextProvider;
      const sectionToAdd = new Chell1DSection('B' as SECONDARY_STRUCTURE_KEYS, 0, 5);

      context.addSelectedSecondaryStructure(sectionToAdd);
      expect(instance.state.selectedSecondaryStructures).toEqual([sectionToAdd]);
    });

    it('Should handle adding a hovered secondary structure.', () => {
      const wrapper = mount(MockContextProvider);
      const context = (MockWrappedElement.props as IMockContextConsumerProps).secondaryStructureContext;
      const instance = wrapper.instance() as SecondaryStructureContextProvider;
      const sectionToAdd = new Chell1DSection('C' as SECONDARY_STRUCTURE_KEYS, 4, 5);

      context.addHoveredSecondaryStructure(sectionToAdd);
      expect(instance.state.hoveredSecondaryStructures).toEqual([sectionToAdd]);
    });

    it('Should handle removing a selected secondary structure.', () => {
      const wrapper = mount(MockContextProvider);
      const context = (MockWrappedElement.props as IMockContextConsumerProps).secondaryStructureContext;
      const instance = wrapper.instance() as SecondaryStructureContextProvider;
      const sectionToAdd = new Chell1DSection('B' as SECONDARY_STRUCTURE_KEYS, 0, 5);

      context.addSelectedSecondaryStructure(sectionToAdd);
      expect(instance.state.selectedSecondaryStructures).toEqual([sectionToAdd]);
      context.removeSecondaryStructure(sectionToAdd);
      expect(instance.state.selectedSecondaryStructures).toEqual([]);
    });

    it('Should remove the hovered and selected secondary structure if removing the selection.', () => {
      const wrapper = mount(MockContextProvider);
      const context = (MockWrappedElement.props as IMockContextConsumerProps).secondaryStructureContext;
      const instance = wrapper.instance() as SecondaryStructureContextProvider;
      const sectionToAdd = new Chell1DSection('B' as SECONDARY_STRUCTURE_KEYS, 0, 5);

      context.addHoveredSecondaryStructure(sectionToAdd);
      context.addSelectedSecondaryStructure(sectionToAdd);
      expect(instance.state.selectedSecondaryStructures).toEqual([sectionToAdd]);
      context.removeSecondaryStructure(sectionToAdd);
      expect(instance.state.hoveredSecondaryStructures).toEqual([]);
      expect(instance.state.selectedSecondaryStructures).toEqual([]);
    });

    it('Should handle removing a hovered secondary structure.', () => {
      const wrapper = mount(MockContextProvider);
      const context = (MockWrappedElement.props as IMockContextConsumerProps).secondaryStructureContext;
      const instance = wrapper.instance() as SecondaryStructureContextProvider;
      const sections = genSecondaryStructureSection('G', 40, 4);

      context.addHoveredSecondaryStructure(sections[0]);
      expect(instance.state.hoveredSecondaryStructures).toEqual(genSecondaryStructureSection('G', 40, 4));
      context.removeHoveredSecondaryStructure(sections[0]);
      expect(instance.state.selectedSecondaryStructures).toEqual([]);
    });

    it('Should not toggle a secondary structure that is already selected.', () => {
      const wrapper = mount(MockContextProvider);
      const context = (MockWrappedElement.props as IMockContextConsumerProps).secondaryStructureContext;
      const instance = wrapper.instance() as SecondaryStructureContextProvider;
      const sections = genSecondaryStructureSection('B', 10, 3);

      context.addSelectedSecondaryStructure(sections[0]);
      context.addHoveredSecondaryStructure(sections[0]);
      expect(instance.state.selectedSecondaryStructures).toEqual(sections);
      context.addHoveredSecondaryStructure(sections[0]);
      expect(instance.state.selectedSecondaryStructures).toEqual(sections);
    });

    it('Should handle removing all secondary structures.', () => {
      const wrapper = mount(MockContextProvider);
      const context = (MockWrappedElement.props as IMockContextConsumerProps).secondaryStructureContext;
      const instance = wrapper.instance() as SecondaryStructureContextProvider;
      const sections = genSecondaryStructureSection('C', 12, 15);

      context.addSelectedSecondaryStructure(sections[0]);
      context.addHoveredSecondaryStructure(sections[0]);
      expect(instance.state.selectedSecondaryStructures).toEqual(sections);
      expect(instance.state.hoveredSecondaryStructures).toEqual(sections);
      context.clearAllSecondaryStructures();
      expect(instance.state.selectedSecondaryStructures).toEqual([]);
      expect(instance.state.hoveredSecondaryStructures).toEqual([]);
    });
  });
});
