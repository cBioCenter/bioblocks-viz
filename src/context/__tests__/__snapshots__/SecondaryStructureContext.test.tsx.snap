// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondaryStructureContext Consumer Should allow components to consume a secondary structure context. 1`] = `
<MockContextClass
  secondaryStructureContext={
    Object {
      "addHoveredSecondaryStructure": [Function],
      "addSelectedSecondaryStructure": [Function],
      "clearAllSecondaryStructures": [Function],
      "hoveredSecondaryStructures": Array [],
      "removeHoveredSecondaryStructure": [Function],
      "removeSecondaryStructure": [Function],
      "selectedSecondaryStructures": Array [],
    }
  }
>
  <div>
    I am a class for a mock context!
  </div>
</MockContextClass>
`;

exports[`SecondaryStructureContext Provider Should allow components to be provided a secondary structure context. 1`] = `
<SecondaryStructureContextProvider>
  <MockContextClass
    secondaryStructureContext={
      Object {
        "addHoveredSecondaryStructure": [Function],
        "addSelectedSecondaryStructure": [Function],
        "clearAllSecondaryStructures": [Function],
        "hoveredSecondaryStructures": Array [],
        "removeHoveredSecondaryStructure": [Function],
        "removeSecondaryStructure": [Function],
        "selectedSecondaryStructures": Array [],
      }
    }
  >
    <div>
      I am a class for a mock context!
    </div>
  </MockContextClass>
</SecondaryStructureContextProvider>
`;
